streams: performance improvements for clearBuffer in Writable

mcollina opened the pull request:
This commits removes a function allocation inside the clearBuffer function. Moreover, it adds counting of buffered chunks to instantiate a fixed-sized Array. The performance improvements are in the range
5-50%, depending on the usecase.

Here is gist of the results of the http benchmarks in node on my MacBook Pro Late 2014 (i7, 16GB of RAM): https://gist.github.com/mcollina/5c59c057cb5b138fcd70.

cc @trevnorris @lucamaraschi @nodejs/benchmarking

geek commented:
LGTM

Fishrock123 commented:
cc @nodejs/streams

trevnorris commented on lib/_stream_writable.js:
```diff
@@ -108,6 +108,31 @@ function WritableState(options, stream) {
+    var entry = state.corkedCbs;
+    var cbs = entry.cbs;
+
+    state.corkedCbs = entry.next;
```
Has this property already been set on the object? If not I would suggest placing it, even just `this.corkedCbs = null;` in the constructor. This will help prevent the object map from changing, and aide the optimizing compiler.

mcollina replied:
It is done some line before: https://github.com/nodejs/node/pull/3751/files#diff-1915a7b992a3012b177dd855fa3477e0R116

mafintosh commented:
:+1: from me

lucamaraschi commented:
:+1: LGTM

chrisdickinson commented on lib/_stream_writable.js:
```diff
@@ -362,27 +388,34 @@ function onwriteDrain(stream, state) {
-    });
+
+    if (state.corkedCbs) {
+      state.corkedCbs.next = new CorkCbs(cbs)
```
Can this only ever be two `corkedCbs` long? If we already have `corkedCbs`, and it has `.next` assigned, will we replace it with a new set of callbacks?

thefourtheye replied:
Missing semi colon

mcollina replied:
@chrisdickinson good spot! I planned to write more about this! My understanding is that it is only 2 `corkedCbs` long. Maybe I am wrong, see these lines, the callback can be deferred on the nextTick: https://github.com/mcollina/node/blob/perf-clearBuffer/lib/_stream_writable.js#L360-L364.

There might also be a different way to implement that, maybe having an array with two objects, and a even/odd counter.

chrisdickinson commented on lib/_stream_writable.js:
```diff
@@ -108,6 +108,31 @@ function WritableState(options, stream) {
   this.errorEmitted = false;
+
+  // count buffered requests
+  this.bufferedRequests = 0;
```
Nit: I'd lean towards naming this attribute something like `bufferedRequestLength` or `bufferedRequestCount` just to make it clear to readers that we're tracking a number.

lucamaraschi replied:
+1 for bufferedRequestCount

chrisdickinson commented:
I like the direction this is going! I might be misunderstanding [this bit](https://github.com/nodejs/node/pull/3751#discussion_r44508925) though — is this behavior intended?

thefourtheye commented on lib/_stream_writable.js:
```diff
@@ -362,27 +388,34 @@ function onwriteDrain(stream, state) {
+    if (state.corkedCbs) {
+      state.corkedCbs.next = new CorkCbs(cbs)
+    } else {
+      state.corkedCbs = new CorkCbs(cbs)
```
Missing semi colon

thefourtheye commented on lib/_stream_writable.js:
```diff
@@ -362,27 +388,34 @@ function onwriteDrain(stream, state) {
     // Fast case, write everything using _writev()
-    var buffer = [];
-    var cbs = [];
+    var buffer = new Array(bufferedRequests);
```
Can these two be maintained in an array of objects?

lucamaraschi replied:
Isn't it going to decrease the  performances in looking-up the values?

thefourtheye replied:
@lucamaraschi Might not be. But if we had to modify one of the arrays, then the other array also has to updated.

mcollina replied:
What do you mean? We need to arrays, one to pass to `droWrite`, and one to call after.

thefourtheye replied:
May be like `[{buffer: ..., cb: ...}, {buffer: ..., cb: ...}, ...]`?

chrisdickinson replied:
@thefourtheye Splitting the buffers and callbacks into two separate arrays is necessary to be able to pass all of the buffers to `_writev` separate from their callbacks. In many cases the list of buffers is forwarded to the C++ layer which expects an array of buffers.

thefourtheye replied:
Oh then its okay I guess.

lucamaraschi replied:
I would still go with the implementation of @mcollina

thefourtheye commented on lib/_stream_writable.js:
```diff
@@ -108,6 +108,31 @@ function WritableState(options, stream) {
+}
+
+function CorkCbs(cbs) {
+  this.cbs = cbs
```
Missing semi colons

mcollina commented:
I've renamed the field, changed the little class into `CorkedCbs`, and added the missing semicolons.

@chrisdickinson are you happy with my explanation on the double element? Do you see another (possibly faster) way to implement this?

jasnell commented:
LGTM

chrisdickinson commented:
> @chrisdickinson are you happy with my explanation on the double element? Do you see another (possibly faster) way to implement this?

I _think_ so, though I might suggest adding a comment (or an `assert`) stating that we only ever expect two of these per stream. Thanks for bearing with me!

mcollina commented:
@chrisdickinson I have added an assert and a comment. It should be ok now.

Let me know if it is ok and I'll squash my commits.

jasnell commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/735/

jasnell commented:
CI Looks good. Couple of related failures.

mcollina commented:
@jasnell anything I should look into/fix? It seems there are windows-specific and arm-specific failures.

On a window configuration, test-child-process-fork-regr-gh-2847.js is failing.

On ARM, I can't access https://ci.nodejs.org/job/node-test-commit-arm/1188/nodes=armv7-wheezy/tapTestReport/ (I get a Gateway Timeout) - so I can't really know.

On ARM fanned (what does that mean?), I'm getting a timeout/failure on test-crypto-dh.js.

jasnell commented:
There's nothing you need to do. The issues are with the CI environment
itself unfortunately
On Nov 15, 2015 2:22 AM, "Matteo Collina" <notifications@github.com> wrote:

> @jasnell <https://github.com/jasnell> anything I should look into/fix? It
> seems there are windows-specific and arm-specific failures.
>
> On a window configuration, test-child-process-fork-regr-gh-2847
> <https://github.com/nodejs/node/issues/2847>.js is failing.
>
> On ARM, I can't access
> https://ci.nodejs.org/job/node-test-commit-arm/1188/nodes=armv7-wheezy/tapTestReport/
> (I get a Gateway Timeout) - so I can't really know.
>
> On ARM fanned (what does that mean?), I'm getting a timeout/failure on
> test-crypto-dh.js.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nodejs/node/pull/3751#issuecomment-156799178>.
>

chrisdickinson commented:
LGTM — I don't have access to the CI server at the moment, perhaps @jasnell can restart CI?

TheAlphaNerd commented:
https://ci.nodejs.org/job/node-test-pull-request/901/

rvagg commented:
@chrisdickinson:
> I don't have access to the CI server at the moment
Do you mean _technically_ or you're on your phone or something that makes it impractical? If you technically can't access it then we should sort that out!

chrisdickinson commented:
@rvagg: ah! I take it back. Last I checked I was in a weird state where I could _kind of_ log in using the old pw auth, but it seems to have resolved. False alarm!

mcollina commented:
I've run this on the CI, and I'm getting a possibly unrelated failure on node-test-binary-windows: https://ci.nodejs.org/job/node-test-binary-windows/301/.

(I've rebased this to master, and squashed the commits)

@chrisdickinson @jasnell if you folks are good with this, I would like to get this merged at some point this week.

cjihrig commented:
Those are known flakey tests, so LGTM.

chrisdickinson commented:
Yep, LGTM — Feel free to merge at will!

mcollina commented:
I've been working on this for the last two days :D.

Turn out V8 is improved in the latest update, and this is not needed anymore in this current form. I am so happy to be proven wrong, because it means all the node apps will be _really faster_ soon, for free.

However there is still the original use case I discovered this for (100+ chunks for a single `_writev`) that is not covered, plus a general cleanup of that area.

I will submit a separate PR for that, referencing this one.
